as-ps/as_ps/main.py

```diff
@@ -100,64 +110,99 @@ def write(self, reason, value):
+            # messages or unnecessary delays. Whether we should extend
+            # it to all power supplies remains to be checked.
+            if App._setpoint_regexp.match(reason):
+                # Accept *-SP and *-Sel right away (not *-Cmd !)
+                self.driver.setParam(reason, value)
+                self.driver.updatePV(reason)
             bbb = self._bbb_devices[pvname.device_name]
-            op = (bbb.write, (pvname.device_name, pvname.propty, value))
+            op = (self._write_operation, (bbb, pvname, value))
             self._prucqueue.append(op)
             self._prucqueue.process()
-            # bbb.write(pvname.device_name, pvname.propty, value)
         else:
+            self.driver.setParam(reason, value)
+            self.driver.updatePV(reason)
             bbb = self._bbb_devices[pvname.device_name]
+            t0 = _time.time()
             bbb.write(pvname.device_name, pvname.propty, value)
-        # return True
+            t1 = _time.time()
+            self._scan_device(bbb, pvname.device_name, force_update=True)
+            _log.info("[{:.2s}] - {:.32s} = {:.50s}".format(
+                'T ', reason, '{:.3f} ms'.format((t1-t0)*1000)))
 
     # --- private methods ---
 
-    def _check_value_changed(self, reason, new_value):
-        # TODO: check how arrays are being compared
-        old_value = self.driver.getParam(reason)
-        if isinstance(new_value, _np.ndarray):
-            return True
-        else:
-            if new_value != old_value:
-                return True
-        return False
+    def _create_bbb_dev_dict(self):
+        # build _bbb_devices dict
+        self._bbb_devices = dict()
+        self._interval = float('Inf')
+        for bbb in self.bbblist:
+            # get minimum time interval for BBB
+            self._interval = min(self._interval, bbb.update_interval())
+            # create bbb_device dict
+            for psname in bbb.psnames:
+                if _use_write_queue and psname.startswith('TB-'):
+                    if self._prucqueue is None:
+                        self._prucqueue = _DequeThread()
+                self._bbb_devices[psname] = bbb
 
-    def _update_ioc_database(self, bbb, device_name):
-        # Return dict indexed with reason
-        data, updated = bbb.read(device_name)
-        if not updated:
-            return
-        for reason, new_value in data.items():
-            if self._check_value_changed(reason, new_value):
-                if new_value is None:
-                    continue
-                self.driver.setParam(reason, new_value)
-            self.driver.setParamStatus(
-                reason, _Alarm.NO_ALARM, _Severity.NO_ALARM)
+    def _write_operation(self, bbb, pvname, value):
+        t0 = _time.time()
+        bbb.write(pvname.device_name, pvname.propty, value)
+        # NOTE: This scan_device might be redundent
+        self._scan_device(bbb, pvname.device_name, force_update=True)
```

Comment:
I recommend you removing this call to `self._scan_device` if it is really not necessary...

Reply:
commented out.